hetmech/dwpc.py

```diff
@@ -0,0 +1,127 @@
+import functools
+import operator
+
+import numpy
+
+from .matrix import metaedge_to_adjacency_matrix, normalize, copy_array
+
+
+def remove_diag(mat):
+    """Set the main diagonal of a square matrix to zeros."""
+    assert mat.shape[0] == mat.shape[1]  # must be square
+    return mat - numpy.diag(mat.diagonal())
```

Comment:
When the underlying matrix is large this is inefficient; the diagonal matrix should be sparse.